In [1]:
# Defining the Required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

# EXPLORATORY DATA ANALYSIS
## Data Understanding, Cleaning and Preparation

In [2]:
# Load the data
data = pd.read_csv('Leads.csv')

### Initial data overview

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

In [4]:
data.describe()

Lead Number    Converted  TotalVisits  Total Time Spent on Website  \
count    9240.000000  9240.000000  9103.000000                  9240.000000   
mean   617188.435606     0.385390     3.445238                   487.698268   
std     23405.995698     0.486714     4.854853                   548.021466   
min    579533.000000     0.000000     0.000000                     0.000000   
25%    596484.500000     0.000000     1.000000                    12.000000   
50%    615479.000000     0.000000     3.000000                   248.000000   
75%    637387.250000     1.000000     5.000000                   936.000000   
max    660737.000000     1.000000   251.000000                  2272.000000   

       Page Views Per Visit  Asymmetrique Activity Score  \
count           9103.000000                  5022.000000   
mean               2.362820                    14.306252   
std                2.161418                     1.386694   
min                0.000000                     7.000000   
25%                1.000000                    14.000000   
50%                2.000000                    14.000000   
75%                3.000000                    15.000000   
max               55.000000                    18.000000   

       Asymmetrique Profile Score  
count                 5022.000000  
mean                    16.344883  
std                      1.811395  
min                     11.000000  
25%                     15.000000  
50%                     16.000000  
75%                     18.000000  
max                     20.000000

In [23]:
data.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   
3                          305                   1.0  ...   
4                         1428                   1.0  ...   

  Digital Advertisement Through Recommendations  \
0                    No                      No   
1                    No                      No   
2                    No                      No   
3                    No                      No   
4                    No                      No   

  Receive More Updates About Our Courses                                 Tags  \
0                                     No          Interested in other courses   
1                                     No                              Ringing   
2                                     No  Will revert after reading the email   
3                                     No                              Ringing   
4                                     No  Will revert after reading the email   

  Update me on Supply Chain Content Get updates on DM Content    City  \
0                                No                        No  Select   
1                                No                        No  Select   
2                                No                        No  Mumbai   
3                                No                        No  Mumbai   
4                                No                        No  Mumbai   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   
3                                       No   
4                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  
3                                     No              Modified  
4                                     No              Modified  

[5 rows x 30 columns]

In [25]:
data.shape

(9240, 30)

In [6]:
# Detailed analysis of missing values
missing_values = data.isnull().sum()
missing_percentage = (missing_values / len(data)) * 100
# Combine into a DataFrame
missing_data = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage': missing_percentage
})

# Display only rows where there are missing values
missing_data = missing_data[missing_data['Missing Values'] > 0]
print(missing_data)

                                               Missing Values  Percentage
Lead Source                                                36    0.389610
TotalVisits                                               137    1.482684
Page Views Per Visit                                      137    1.482684
Last Activity                                             103    1.114719
Country                                                  2461   26.634199
Specialization                                           1438   15.562771
How did you hear about X Education                       2207   23.885281
What is your current occupation                          2690   29.112554
What matters most to you in choosing a course            2709   29.318182
Tags                                                     3353   36.287879
Lead Quality                                             4767   51.590909
Lead Profile                                             2709   29.318182
City                                  

In [7]:
# Identify categorical variables with 'Select' as a placeholder for missing data
select_as_missing = (data == 'Select').sum()
print("Count of 'Select' as placeholder:\n", select_as_missing)

Count of 'Select' as placeholder:
 Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                         0
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                         0
Total Time Spent on Website                         0
Page Views Per Visit                                0
Last Activity                                       0
Country                                             0
Specialization                                   1942
How did you hear about X Education               5043
What is your current occupation                     0
What matters most to you in choosing a course       0
Search                                              0
Magazine                                       

In [13]:
# Select and Missing data

select_missing_data = select_as_missing+missing_values
select_missing_per = (select_missing_data/ len(data)) * 100
# Combine into a DataFrame
Select_missing_data = pd.DataFrame({
    'Sel & Miss Values': select_missing_data,
    'Percentage': select_missing_per
})

# Display only rows where there are missing values
Select_missing_data = Select_missing_data[Select_missing_data['Sel & Miss Values'] > 0]
print(Select_missing_data.sort_values('Percentage'))

                                               Sel & Miss Values  Percentage
Lead Source                                                   36    0.389610
Last Activity                                                103    1.114719
TotalVisits                                                  137    1.482684
Page Views Per Visit                                         137    1.482684
Country                                                     2461   26.634199
What is your current occupation                             2690   29.112554
What matters most to you in choosing a course               2709   29.318182
Tags                                                        3353   36.287879
Specialization                                              3380   36.580087
City                                                        3669   39.707792
Asymmetrique Activity Score                                 4218   45.649351
Asymmetrique Activity Index                                 4218   45.649351

In [9]:
# Load the data dictionary
data_dict = pd.read_excel('Leads Data Dictionary.xlsx', header=2, usecols='B:C')

In [10]:
data_dict.head(40)

Variables  \
0                                        Prospect ID   
1                                        Lead Number   
2                                        Lead Origin   
3                                        Lead Source   
4                                       Do Not Email   
5                                        Do Not Call   
6                                          Converted   
7                                        TotalVisits   
8                        Total Time Spent on Website   
9                               Page Views Per Visit   
10                                     Last Activity   
11                                           Country   
12                                    Specialization   
13                How did you hear about X Education   
14                   What is your current occupation   
15  What matters most to you in choosing this course   
16                                            Search   
17                                          Magazine   
18                                 Newspaper Article   
19                                X Education Forums   
20                                         Newspaper   
21                             Digital Advertisement   
22                           Through Recommendations   
23            Receive More Updates About Our Courses   
24                                              Tags   
25                                      Lead Quality   
26                 Update me on Supply Chain Content   
27                         Get updates on DM Content   
28                                      Lead Profile   
29                                              City   
30                       Asymmetrique Activity Index   
31                        Asymmetrique Profile Index   
32                       Asymmetrique Activity Score   
33                        Asymmetrique Profile Score   
34          I agree to pay the amount through cheque   
35            a free copy of Mastering The Interview   
36                             Last Notable Activity   

                                          Description  
0   A unique ID with which the customer is identif...  
1       A lead number assigned to each lead procured.  
2   The origin identifier with which the customer ...  
3   The source of the lead. Includes Google, Organ...  
4   An indicator variable selected by the customer...  
5   An indicator variable selected by the customer...  
6   The target variable. Indicates whether a lead ...  
7   The total number of visits made by the custome...  
8   The total time spent by the customer on the we...  
9   Average number of pages on the website viewed ...  
10  Last activity performed by the customer. Inclu...  
11                       The country of the customer.  
12  The industry domain in which the customer work...  
13  The source from which the customer heard about...  
14  Indicates whether the customer is a student, u...  
15  An option selected by the customer indicating ...  
16  Indicating whether the customer had seen the a...  
17                                                NaN  
18                                                NaN  
19                                                NaN  
20                                                NaN  
21                                                NaN  
22  Indicates whether the customer came in through...  
23  Indicates whether the customer chose to receiv...  
24  Tags assigned to customers indicating the curr...  
25  Indicates the quality of lead based on the dat...  
26  Indicates whether the customer wants updates o...  
27  Indicates whether the customer wants updates o...  
28  A lead level assigned to each customer based o...  
29                          The city of the customer.  
30  An index and score assigned to each customer b...  
31                                                NaN  
32                                                NaN  
33        

## Handling Missing Data

#### The following columns have a high number of missing or select place holder infomation
    Field | Missing /Select Values | Percentage Missing
    1. Tags                                                        3353   36.287879
    2. Specialization                                              3380   36.580087
    3. City                                                        3669   39.707792
    4. Asymmetrique Activity Score                                 4218   45.649351
    5. Asymmetrique Activity Index                                 4218   45.649351
    6. Asymmetrique Profile Index                                  4218   45.649351
    7. Asymmetrique Profile Score                                  4218   45.649351
    8. Lead Quality                                                4767   51.590909
    9. Lead Profile                                                6855   74.188312
    10. How did you hear about X Education                          7250   78.463203

##### Dropping the column with very high missing values (>40%)


In [14]:
data.drop(['How did you hear about X Education'], axis=1, inplace=True)

In [15]:
data.drop(['Lead Profile'], axis=1, inplace=True)

In [16]:
data.drop(['Lead Quality'], axis=1, inplace=True)

In [17]:
data.drop(['Asymmetrique Profile Score'], axis=1, inplace=True)

In [18]:
data.drop(['Asymmetrique Profile Index'], axis=1, inplace=True)

In [19]:
data.drop(['Asymmetrique Activity Index'], axis=1, inplace=True)

In [20]:
data.drop(['Asymmetrique Activity Score'], axis=1, inplace=True)

In [22]:
# Reassessing the shape of the data set
data.shape

(9240, 30)

##### Missing data handling
The following still consists of missing information up to 39%
    0. Field                                           Sel & Miss Values  Percentage
    1. Lead Source                                                   36    0.389610
    2. Last Activity                                                103    1.114719
    4. TotalVisits                                                  137    1.482684
    5. Page Views Per Visit                                         137    1.482684
    6. Country                                                     2461   26.634199
    7. What is your current occupation                             2690   29.112554
    8. What matters most to you in choosing a course               2709   29.318182
    9. Tags                                                        3353   36.287879
    10. Specialization                                              3380   36.580087
    11. City                                                        3669   39.707792

This is a lot of information and removing the respective columns or rows cannot be determined at this time. 
To better handle this, all "NA" and "Select" value will ne replaced with "Not Provided" to make it easy to 
identify during data analysis and handle in case of exception

In [26]:
# Replace missing values and 'Select' with 'Not Provided'
columns_with_missing_data = [
    'Lead Source', 'Last Activity', 'TotalVisits', 'Page Views Per Visit',
    'Country', 'What is your current occupation', 'What matters most to you in choosing a course',
    'Tags', 'Specialization', 'City'
]

for column in columns_with_missing_data:
    data[column] = data[column].fillna('Not Provided').replace('Select', 'Not Provided')

In [27]:
data.shape

(9240, 30)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 30 columns):
 #   Column                                         Non-Null Count  Dtype 
---  ------                                         --------------  ----- 
 0   Prospect ID                                    9240 non-null   object
 1   Lead Number                                    9240 non-null   int64 
 2   Lead Origin                                    9240 non-null   object
 3   Lead Source                                    9240 non-null   object
 4   Do Not Email                                   9240 non-null   object
 5   Do Not Call                                    9240 non-null   object
 6   Converted                                      9240 non-null   int64 
 7   TotalVisits                                    9240 non-null   object
 8   Total Time Spent on Website                    9240 non-null   int64 
 9   Page Views Per Visit                           9240 non-null   